vimiv/utils/imageheader.py

```diff
@@ -0,0 +1,411 @@
+| XPM    | xpm                       | yes                      |
+| SVG    | svg                       | yes                      |
+| SVG    | svgz                      | no (different from svg?) |
+| ---    | ---                       | ---                      |
+
+Extended QT Support:
+| ---    | ---                       | ---                        |
+| Format | Extension according to QT | Vimiv Supported            |
+| ---    | ---                       | ---                        |
+| ICNS   | ico                       | yes                        |
+| ICNS   | icns                      | yes                        |
+| ICNS   | cur                       | no (what are magic bytes?) |
+| JP2    | TODO: ?                   | yes                        |
+| MNG    | TODO: ?                   | no (what are magic bytes?) |
+| TGA    | tga                       | no (undetectable)          |
+| TIFF   | tif, tiff                 | yes                        |
+| WBMP   | wbmp                      | no (what are magic bytes?) |
+| WEBP   | webp                      | yes                        |
+| ---    | ---                       | ---                        |
+
+A great list of magic bytes is provided here:
+https://en.wikipedia.org/wiki/List_of_file_signatures
+"""
+
+import contextlib
+import functools
+
+from typing import Optional, List, Callable, Any, Tuple
+
+from PyQt5.QtGui import QImageReader
+
+from vimiv.utils import log, imagereader
+
+_logger = log.module_logger(__name__)
+
+# List containing all registered check functions
+_registry: List[Tuple[str, Callable[[bytes], bool]]] = []
+
+
+def detect(filename: str) -> Optional[str]:
+    """Determine type of image based on the magic bytes.
+
+    Evaluates each registered check function in the order they were registered. If
+    registered with `priority`, then that check is evaluated before all checks with out
+    `priority`.
+
+    Args:
+        filename: Name of file to determine type of.
+
+    Returns:
+        Filetype or None if unknown.
+    """
+    with open(filename, "rb") as f:
+        header = f.read(32)
+
+    for filetype, check in _registry:
+        with contextlib.suppress(IndexError):
+            if check(header):
+                return filetype
+    return None
+
+
+def register(
+    filetype: str, check: Any, priority: bool = False, validate: bool = True
+) -> None:
+    """Register format test function.
+
+    Args:
+        filetype: Name of the format to register.
+        check: Test function for that type.
+        priority: Evaluate check before all previously registered checks if true.
+        validate: Validate that the type is supported at runtime.
+    """
+
+    @functools.wraps(check)
+    def check_verified(header: bytes) -> bool:
+        """Tests at runtime whether a filetype is actually supported.
+
+        If not, then the filetype check is unregistered.
+        """
+        if not validate:
```

Comment:
I think it would be cleaner to move this check out of the actual function, i.e. register `check` directly instead of `check_verified` later, and avoid having this if in every function call.

Reply:
Makes totally sense. Simply doing:
```
    if validate:
        check = check_verified

    if priority:
        _registry.insert(0, (filetype, check))
    else:
        _registry.append((filetype, check))
```

does unfortunatelly result in some recusion issue (see testcases). Or do I miss something obvious?

Pushed a less clean solution: 7af4fbea94bcd18cc676469b20e86a04f5f32ddf

Reply:
The initial option probably creates some type of recursion havoc, as the actual `check_verified` body hasn't been called yet.

How about using
```
    check_register = check_verified if validate else check
```
to make it a bit easier to parse?

For some reason this also fixes the mypy error - win-win :blush:

Reply:
My brain must have been to confused about the recursion weirdness that I did not think about the ternary operator 🙈 Fixed in 63f8c954b4b4600453b79c9f9cd75b15c3676cf7